R/data_read.R

```diff
@@ -106,40 +118,56 @@ data_read <- function(path, path_catalog = NULL, encoding = NULL, verbose = TRUE
 
 # process imported data from SPSS, SAS or Stata -----------------------
 
-.post_process_imported_data <- function(x, verbose) {
-  if (verbose) {
-    message("Preparing data... Almost there!")
-  }
-  x[] <- lapply(x, function(i) {
-    # save labels
-    value_labels <- attr(i, "labels", exact = TRUE)
-    variable_labels <- attr(i, "label", exact = TRUE)
-
-    # filter, so only matching value labels remain
-    value_labels <- value_labels[value_labels %in% unique(i)]
-
-    # guess variable type
-    if (!is.character(i)) {
-      # if all values are labelled, we assume factor. Use labels as levels
-      if (length(value_labels) == insight::n_unique(i)) {
-        i <- factor(as.character(i), labels = names(value_labels))
-        value_labels <- NULL
-      } else {
-        i <- as.numeric(i)
-      }
-    } else {
-      i <- as.character(i)
-    }
-
-    # drop unused value labels
-    if (!is.null(value_labels) && length(value_labels <- value_labels[value_labels %in% unique(i)])) {
-      attr(i, "labels") <- value_labels
+.post_process_imported_data <- function(x, convert_factors, verbose) {
+  # user may decide whether we automatically detect variable type or not
+  if (isTRUE(convert_factors)) {
+    if (verbose) {
+      message("Preparing data... Almost there!")
     }
-
-    # add back variable label
-    attr(i, "label") <- variable_labels
-    i
-  })
+    x[] <- lapply(x, function(i) {
+      # only proceed if not all missing
+      if (!all(is.na(i))) {
+        # save labels
+        value_labels <- attr(i, "labels", exact = TRUE)
+        variable_labels <- attr(i, "label", exact = TRUE)
+
+        # filter, so only matching value labels remain
+        value_labels <- value_labels[value_labels %in% unique(i)]
+
+        # guess variable type
+        if (!is.character(i)) {
+          # if all values are labelled, we assume factor. Use labels as levels
+          if (!is.null(value_labels) && length(value_labels) == insight::n_unique(i)) {
+            # convert into factor - this usually preserves numeric levels
+            # when input variable was numeric
+            i <- factor(as.character(i))
+            lvl <- tryCatch(
+              # if we have numeric values, make sure we match order of labels
+              names(value_labels[match(as.numeric(levels(i)), value_labels)]),
+              # if anything fails, use order of labels as levels
+              warning = function(w) names(value_labels),
+              error = function(e) names(value_labels)
+            )
+            try(levels(i) <- lvl, silent = TRUE)
+            value_labels <- NULL
+          } else {
+            i <- as.numeric(i)
+          }
+        } else {
+          i <- as.character(i)
```

Comment:
this `else` only applies if `!is.character(i)` is FALSE so I don't think it's needed